support/ebpf/tracemgmt.h

```diff
@@ -444,7 +444,213 @@ void tail_call(void *ctx, int next) {
+// without valid user mode pt_regs so is_kernel_address(pc) is not enough.
+static inline
+bool ptregs_is_usermode(struct pt_regs *regs) {
+#if defined(__x86_64__)
+  // On x86_64 the user mode SS should always be __USER_DS.
+  if (regs->ss != __USER_DS) {
+    return false;
+  }
+  return true;
+#elif defined(__aarch64__)
+  // Check if the processor state is in the EL0t what linux uses for usermode.
+  if ((regs->pstate & PSR_MODE_MASK) != PSR_MODE_EL0t) {
+    return false;
+  }
+  return true;
+#else
+#error add support for new architecture
+#endif
+}
+
+// Extract the usermode pt_regs for current task. Use context given pt_regs
+// if it is usermode regs, or resolve it via struct task_struct.
+//
+// State registers are not touched (get_pristine_per_cpu_record already reset it)
+// if something fails. has_usermode_regs is set to true if a user-mode register
+// context was found: not every thread that we interrupt will actually have
+// a user-mode context (e.g. kernel worker threads won't).
+static inline ErrorCode get_usermode_regs(struct pt_regs *ctx,
+                                          UnwindState *state,
+                                          bool *has_usermode_regs) {
+  ErrorCode error;
+
+  if (!ptregs_is_usermode(ctx)) {
+    u32 key = 0;
+    SystemConfig* syscfg = bpf_map_lookup_elem(&system_config, &key);
+    if (!syscfg) {
+      // Unreachable: array maps are always fully initialized.
+      return ERR_UNREACHABLE;
+    }
+
+    // Use the current task's entry pt_regs
+    struct task_struct *task = (struct task_struct *) bpf_get_current_task();
+    long ptregs_addr = get_task_pt_regs(task, syscfg);
+
+    struct pt_regs regs;
+    if (!ptregs_addr || bpf_probe_read_kernel(&regs, sizeof(regs), (void*) ptregs_addr)) {
+      increment_metric(metricID_UnwindNativeErrReadKernelModeRegs);
+      return ERR_NATIVE_READ_KERNELMODE_REGS;
+    }
+
+    if (!ptregs_is_usermode(&regs)) {
+      // No usermode registers context found.
+      return ERR_OK;
+    }
+    error = copy_state_regs(state, &regs, true);
+  } else {
+    // User mode code interrupted, registers are available via the ebpf context.
+    error = copy_state_regs(state, ctx, false);
+  }
+  if (error == ERR_OK) {
+    DEBUG_PRINT("Read regs: pc: %llx sp: %llx fp: %llx", state->pc, state->sp, state->fp);
+    *has_usermode_regs = true;
+  }
+  return error;
+}
+
+#else // TESTING_COREDUMP
+
+static inline ErrorCode get_usermode_regs(struct pt_regs *ctx,
+                                          UnwindState *state,
+                                          bool *has_usermode_regs) {
+  // Coredumps provide always usermode pt_regs directly.
+  ErrorCode error = copy_state_regs(state, ctx, false);
+  if (error == ERR_OK) {
+    *has_usermode_regs = true;
+  }
+  return error;
+}
+
+#endif // TESTING_COREDUMP
+
+static inline
+int collect_trace(struct pt_regs *ctx, TraceOrigin origin, u32 pid, u32 tid, u64 off_cpu_time) {
```

Comment:
With moving `collect_trace()` to `tracemgmt.h` its arguments changed. `origin` was added to further keep track what triggered the stack unwinding. `pid` and `tid` also got added as argument, as the entry point for off CPU profiling does some filtering on these parameters, so it didn't make sense to call the helper for this information multiple times. And finally `off_cpu_time` got added as argument, to forward the information for how long a trace was off CPU and store the information in the `Trace` struct.